src/utils/nanobench.py

```diff
@@ -0,0 +1,129 @@
+        total_ms += (end - start).total_seconds() * 1000
+        print
+    if expected_resp is None or "error" in json.loads(expected_resp):
+        fileno = "[ERRORED]"
+    else:
+        fileno = len(json.loads(expected_resp)["result"]["files"])
+    print(f"{query:<60} average {total_ms/TRY_COUNT: 10.3f} files: {fileno}")
+
+
+def nano(socket):
+    """ Simple queries. They should work on any database """
+    benchmark(socket, 'select "abc";')
+    benchmark(socket, 'select "abcdefgh";')
+    benchmark(socket, 'select "abc" & "qwe" & "zxc";')
+    benchmark(socket, 'select "abc" | "qwe" | "zxc";')
+    benchmark(socket, 'select min 1 of ("abc", "qwe", "zxc");')
+    benchmark(socket, 'select min 2 of ("abc", "qwe", "zxc");')
+    benchmark(socket, 'select min 3 of ("abc", "qwe", "zxc");')
+    benchmark(socket, "select {61 62 6?};")
+    benchmark(socket, "select {61 62 6? 63};")
+    benchmark(socket, "select {61 62 6? 63 64};")
+
+
+def mini(socket):
+    """ Reasonable queries. They may take some time, but should return results
+    in a reasonable time. """
+    benchmark(socket, "select {60 61 62 ??};")
+    benchmark(socket, "select {60 61 62 ?? 63};")
+    benchmark(socket, "select {60 61 62 ?? 63 64};")
+    benchmark(socket, "select {61 62 ??};")
+    benchmark(socket, "select {61 62 ?? 63};")
+    benchmark(socket, "select {61 62 ?? 63 64};")
+    benchmark(socket, "select {(61|62) (62|63) (64|65)};")
+    benchmark(socket, "select {(61|62) (62|63) (64|65) (66|67)};")
+    benchmark(socket, "select {(61|62) (62|63) (64|65) (66|67) (68|69)};")
+
+
+def heavyduty(socket):
+    """ Heavy queries. Used to benchmark querygraphs. Most of them will take
+    forever on a large real-world database. """
+    benchmark(socket, "select {?1 ?2 ?3};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5 ?6};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5 ?6 ?7};")
+    benchmark(socket, "select {?1 ?2 ?3 ?? ?5 ?6 ?7};")
+    benchmark(socket, "select {62 ?? 63 ?? 64};")
+    benchmark(socket, "select {1? 2? ?? 1? 2?};")
+    benchmark(socket, "select {61 62 ?? 63 ?? 64 65};")
+    benchmark(socket, "select {61 62 ?? 63 ?? 64 ?? 65 66};")
+    benchmark(socket, "select {61 62 63 64 64 1? 2? 3? 4? 5?};")
+    benchmark(socket, "select {1? 2? 3? 4? 5? 61 62 63 64 64};")
+    benchmark(socket, "select {61 62 63 64 ?? ??};")
+    benchmark(socket, "select {?? ?? 61 62 63 64};")
+    benchmark(socket, "select {61 62 63 64 ?? ?? 65};")
+    benchmark(socket, "select {65 ?? ?? 61 62 63 64};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5 ?6};")
+    benchmark(socket, "select {61 62 63} & {?1 ?2 ?3 ?4 ?5 ?6};")
+    benchmark(socket, "select {61 62 63 ?? ?? ?? ?1 ?2 ?3 ?4 ?5 ?6 ?8};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5 ?6} & {61 62 63};")
+    benchmark(socket, "select {?1 ?2 ?3 ?4 ?5 ?6 ?? ?? ?? 61 62 63};")
+
+
+def main() -> None:
```

Comment:
see `ursadb_bench` command in ursadb - this is similar functionality

Reply:
Not really though - ursadb_bench benchmarks `index+merge`, while this tests for query time.

But in the future I think `ursadb_bench` will get a rewrite in python too, just not now (e2e tests already do the same thing, but they don't measure time)

Reply:
Another advantage that I have in mind, that having this in an independent script will make it easy to benchmark all ursadb versions in a fair way - i.e. do some git magic to compile every ursadb version in existence, run benchmark on it and graph results. That sounds like a cool idea to see preformance improvements/regressions over the project's history.